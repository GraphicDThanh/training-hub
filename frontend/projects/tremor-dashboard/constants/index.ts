export * from "./account";
export * from "./analytics";
export * from "./api";
export * from "./billing";
export * from "./breakpoints";
export * from "./colors";
export * from "./common";
export * from "./flagSrc";
export * from "./images";
export * from "./pagination";
export * from "./messages";
export * from "./pincode";
export * from "./products";
export * from "./projects";
export * from "./profile";
export * from "./regex";
export * from "./routes";
export * from "./sales";
export * from "./saleStatistical";
export * from "./sections";
export * from "./status";
export * from "./units";
export * from "./keyboard";
export * from "./steps";
export * from "./users";
export * from "./conversation";
