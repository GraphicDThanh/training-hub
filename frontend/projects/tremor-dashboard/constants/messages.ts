export const MESSAGES_ERROR = {
  NAME_MIN_LENGTH: "Minimum length should be 4.",
  NAME_INVALID: "This field is no space at start or end",
  NAME_REQUIRED: "Name is required.",
  PASSWORD_REQUIRED: "Password is required.",
  EMAIL_REQUIRED: "Email is required.",
  EMAIL_INVALID: "Invalid email address.",
  SKU_INVALID: "Invalid SKU number.",
  PASSWORD_INVALID:
    "Password must have at least 8 characters, 1 special character and 1 number.",
  FIELD_REQUIRED: "This field is required.",
  MIN_LENGTH_4: "This field must have more than 4 characters.",
  URL_REQUIRED: "URL is required.",
  INVALID_URL: "URL is invalid.",
  INVALID_PHONE_NUMBER: "Phone number is invalid",
  INVALID_BIRTH_DATE: "Birth date is invalid.",
  REPEAT_PASSWORD_INVALID: "Password not matching",
  TERMS_REQUIRED: "Please agree with the terms and conditions to continue.",
  CONFIRM_EMAIL: "The email confirmation does not match.",
  PRODUCT: {
    WEIGHT_LIMIT: "Weight has limit is 2000 pound.",
    QUANTITY_LIMIT: "Quantity has limit is 10000.",
    PRICE_LIMIT: "Price has limit is $20,000.",
  },
  TRANSACTION: {
    AMOUNT_LIMIT: "Amount has limit is $20,000.",
    MIN_AMOUNT: "Amount must be positive number.",
  },
};

export const MESSAGES_WARNING = {
  PRODUCT: {
    WEIGHT_IS_ZERO: "Missing weight is not encourage.",
    QUANTITY_IS_ZERO:
      "This product will has label sold out if have quantity default is 0.",
    ACCEPT_IMAGE_UPLOAD:
      "(We accept .png, .jpg, .jpeg, and .webp files, up to 32MB in size.)",
  },
};

export const MESSAGE_SIGN_UP = {
  SUCCESS: "Sign up successfully.",
  FAILED: "Failed to sign up",
};

export const MESSAGE_SIGN_IN = {
  FAILED: "Failed to sign in",
  SUCCESS: "Sign in successfully.",
  CREDENTIALS: "Please make sure your email address and password are correct.",
};

export const MESSAGE_ADD_PRODUCT = {
  SUCCESS: "Add product successfully",
  FAILED: "Failed to add product",
};

export const MESSAGE_EDIT_PRODUCT = {
  SUCCESS: "Edit product successfully",
  FAILED: "Failed to edit product",
};

export const MESSAGE_USER = {
  MAIL_EXISTS: "Account with this email already exists!",
  ADD_FAILED: "Failed to create new account. Please try again!",
  EDIT_SUCCESS: "Edit user successfully",
  EDIT_FAILED: "Failed to edit user",
  GET_USER_FAILED: "Failed to get user",
};

export const MESSAGE_SIGN_OUT = {
  SUCCESS: "Sign out successfully.",
  FAILED: "Failed to Sign out",
};

export const MESSAGE_ADD_PROJECT = {
  SUCCESS: "Add project successfully",
  FAILED: "Failed to add project",
};

export const MESSAGE_EDIT_PROJECT = {
  SUCCESS: "Edit project successfully",
  FAILED: "Failed to edit project",
};

export const MESSAGE_DELETE_PROJECT = {
  SUCCESS: "Delete project successfully",
  FAILED: "Failed to delete project",
};

export const MESSAGE_ADD_USER = {
  SUCCESS: "Add new user successfully",
  FAILED: "Failed to add new user",
};

export const MESSAGES_TRANSACTION = {
  ADD_SUCCESS: "Create new transaction successfully",
  ADD_FAILED: "Failed to create new transaction. Please try again!",
};
