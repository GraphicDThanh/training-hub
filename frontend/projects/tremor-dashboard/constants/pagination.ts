export const PAGE_SIZE = {
  SIZE: 10,
};
