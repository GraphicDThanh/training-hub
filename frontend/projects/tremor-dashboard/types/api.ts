/**
 * If you want the API to return something for your ease of handling the UI, style that variable here
 */
export interface IResponseFromAPI<Q> {
  errorMessages: Array<Record<string, string>>;
  data: Q | null;
}
