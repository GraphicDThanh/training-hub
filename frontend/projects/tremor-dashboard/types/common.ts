export interface Step {
  index: number;
  content: string;
}

export interface IMedia {
  image: string;
}

export interface CdnResponse {
  data: {
    width: number;
    height: number;
    image: ImageInfo;
  };
}

export interface ImageInfo {
  filename: string;
  url: string;
}

export interface SortItem {
  key: string;
  direction: string;
}

export type TSearchParams = {
  id?: string;
  query?: string;
  isAvailable?: string;
  status?: string;
  page?: number;
  sortBy?: string;
  orderBy?: string;
};

export interface TFieldSocial {
  name: string;
  id: string;
  label: string;
  value: string;
}
