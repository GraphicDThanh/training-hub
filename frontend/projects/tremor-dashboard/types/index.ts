export * from "./analytics";
export * from "./api";
export * from "./billing";
export * from "./breadcrumb";
export * from "./common";
export * from "./invoice";
export * from "./navigation";
export * from "./orderDetails";
export * from "./orders";
export * from "./product";
export * from "./profile";
export * from "./project";
export * from "./sales";
export * from "./select";
export * from "./table-list";
export * from "./table";
export * from "./transactions";
export * from "./user";
