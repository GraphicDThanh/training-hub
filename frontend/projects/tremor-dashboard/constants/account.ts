// Types
import { OptionType } from "@/types";

export const GENDERS: OptionType[] = [
  {
    option: "Male",
    value: "0",
  },
  {
    option: "Female",
    value: "1",
  },
];

export const SKILLS: Record<string, string | number>[] = [
  {
    value: "0",
    option: "React",
  },
  {
    value: "1",
    option: "Vue",
  },
  {
    value: "2",
    option: "Angular",
  },
  {
    value: "3",
    option: "Svelte",
  },
  {
    value: "4",
    option: "Javascript",
  },
];

export const MONTHS: OptionType[] = [
  {
    option: "January",
    value: "0",
  },
  {
    option: "February",
    value: "1",
  },
  {
    option: "March",
    value: "2",
  },
  {
    option: "April",
    value: "3",
  },
  {
    option: "May",
    value: "4",
  },
  {
    option: "June",
    value: "5",
  },
  {
    option: "July",
    value: "6",
  },
  {
    option: "August",
    value: "7",
  },
  {
    option: "September",
    value: "8",
  },
  {
    option: "October",
    value: "9",
  },
  {
    option: "November",
    value: "10",
  },
  {
    option: "December",
    value: "11",
  },
];

export const ROLES = [
  {
    value: 0,
    label: "User",
  },
  {
    value: 1,
    label: "Admin",
  },
];
