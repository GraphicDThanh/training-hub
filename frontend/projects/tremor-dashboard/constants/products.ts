// Types
import { OptionType, TFieldSocial } from "@/types";

type TopSellingProductsType = {
  [key: number]: string;
  1: string;
  2: string;
  3: string;
  4: string;
  5: string;
};

export const TOP_SELLING_PRODUCTS_IMAGE: TopSellingProductsType = {
  1: "/assets/images/products/blue-shoe.webp",
  2: "/assets/images/products/black-mug.webp",
  3: "/assets/images/products/black-chair.webp",
  4: "/assets/images/products/bang-sound.webp",
  5: "/assets/images/products/photo-tools.webp",
};

export const PRICE_TYPE: OptionType[] = [
  {
    option: "BTC",
    value: "1",
  },
  {
    option: "CNY",
    value: "2",
  },
  {
    option: "EUR",
    value: "3",
  },
  {
    option: "GBP",
    value: "4",
  },
  {
    option: "INR",
    value: "5",
  },
  {
    option: "USD",
    value: "6",
  },
];

export const PRICE_TAGS: OptionType[] = [
  {
    option: "Black Friday",
    value: "1",
  },
  {
    option: "Expired",
    value: "2",
  },
  {
    option: "Sale",
    value: "3",
  },
];

export const PRODUCT_CATEGORY: OptionType[] = [
  {
    option: "Clothing",
    value: "1",
  },
  {
    option: "Electronics",
    value: "2",
  },
  {
    option: "Furniture",
    value: "3",
  },
  {
    option: "Others",
    value: "4",
  },
  {
    option: "Real Estate",
    value: "5",
  },
];

export const OTHER_PRODUCTS_LENGTH = 4;

export const SOCIAL_PRODUCTS: TFieldSocial[] = [
  {
    id: "shopifyUrl",
    name: "shopifyUrl",
    label: "Shopify Handle",
    value: "",
  },
  {
    id: "facebookUrl",
    name: "facebookUrl",
    label: "Facebook Account",
    value: "",
  },
  {
    id: "instagramUrl",
    name: "instagramUrl",
    label: "Instagram Account",
    value: "",
  },
];
