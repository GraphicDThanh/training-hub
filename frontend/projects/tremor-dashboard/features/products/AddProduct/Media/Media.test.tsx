import { fireEvent, render } from "@testing-library/react";
import { useForm } from "react-hook-form";
import { ReactNode } from "react";

// Components
import Media from "./Media";

//Types
import { IMedia } from "@/types";

// Constants
import { DRAG_ZONE } from "@/constants";

interface ControllerProps {
  name?: string;
  render?: (arg0: {
    field?: {
      value?: string[] | string;
      name?: string;
      onChange?: jest.Mock<string, string[], string>;
    };
    formState?: {
      field?: { name: string; rules: Object };
      errors?: {};
    };
    fieldState?: { error: { message?: string } };
  }) => ReactNode;
}

const mockControllerNonError = (name: string) => ({
  field: {
    name: name,
    onChange: jest.fn(),
    value: [],
  },
  formState: { errors: {} },
});

jest.mock("react-hook-form", () => ({
  ...jest.requireActual("react-hook-form"),
  // Mock the useFormContext hook
  useForm: jest.fn(() => {
    return {
      formState: {
        errors: {},
      },
    };
  }),
  // Mock the Controller
  Controller: (props: ControllerProps) => {
    return props?.render?.(mockControllerNonError(props.name as string));
  },
}));

const onUploadMock = jest.fn();

describe("Media component", () => {
  const { control } = useForm<IMedia>();

  // Create a PNG file object
  const file = new File(["test image"], "test.png", { type: "image/png" });

  const setupRender = () =>
    render(<Media control={control} onUpload={onUploadMock} />);

  it("Should match snapshot", () => {
    const { container } = setupRender();
    expect(container).toMatchSnapshot();
  });

  it("should call onUpload if user inputs an image file", () => {
    const { getByTestId } = setupRender();

    const inputElement = getByTestId("dropzone-file");

    // Simulate a change event with image file
    fireEvent.change(inputElement, { target: { files: [file] } });

    expect(onUploadMock).toHaveBeenCalledWith(file);
  });

  it("should call onUpload if drop an image that has name match TYPE_FILE_IMAGE_REGEX", () => {
    const { getByTestId } = setupRender();

    const labelElement = getByTestId("drag-file");

    // Simulate a onDrop event with image file
    fireEvent.drop(labelElement, { dataTransfer: { files: [file] } });

    // expect onUpload called
    const inputElement = getByTestId("dropzone-file");

    // Simulate a change event with image file
    fireEvent.change(inputElement, { target: { files: [file] } });

    expect(onUploadMock).toHaveBeenCalledWith(file);
  });

  it("should not call onUpload and update input styles if drop an image that has name not match regex", () => {
    const { getByTestId, getByText } = setupRender();

    const labelElement = getByTestId("drag-file");

    // Expect input, label styles and text before onDrop
    expect(getByText(DRAG_ZONE.DEFAULT.TEXT)).toBeTruthy();
    expect(
      labelElement.className.includes(DRAG_ZONE.DEFAULT.STYLE_INPUT),
    ).toBeTruthy();
    expect(
      labelElement
        .querySelectorAll("p")[0]
        .className.includes(DRAG_ZONE.DEFAULT.STYLE_TEXT),
    ).toBeTruthy();

    // Simulate a onDrop event with empty file
    fireEvent.drop(labelElement, { dataTransfer: { files: [] } });

    // Expect input, label styles change and text after onDrop
    expect(getByText(DRAG_ZONE.ON_WRONG_FORMAT.TEXT)).toBeTruthy();
    expect(
      labelElement.className.includes(DRAG_ZONE.ON_WRONG_FORMAT.STYLE_INPUT),
    ).toBeTruthy();
    expect(
      labelElement
        .querySelectorAll("p")[0]
        .className.includes(DRAG_ZONE.ON_WRONG_FORMAT.STYLE_TEXT),
    ).toBeTruthy();
  });

  it("should update input component correctly if drag image into input", () => {
    const { getByTestId, getByText } = setupRender();

    const labelElement = getByTestId("drag-file");

    // Expect input, label styles and text before dragEnter
    expect(getByText(DRAG_ZONE.DEFAULT.TEXT)).toBeTruthy();
    expect(
      labelElement.className.includes(DRAG_ZONE.DEFAULT.STYLE_INPUT),
    ).toBeTruthy();
    expect(
      labelElement
        .querySelectorAll("p")[0]
        .className.includes(DRAG_ZONE.DEFAULT.STYLE_TEXT),
    ).toBeTruthy();

    // Simulate a dragImage event with the file
    fireEvent.dragEnter(labelElement, { dataTransfer: { files: [file] } });

    // Expect input, label styles change and text after dragEnter
    expect(getByText(DRAG_ZONE.ON_DRAG.TEXT)).toBeTruthy();
    expect(
      labelElement.className.includes(DRAG_ZONE.ON_DRAG.STYLE_INPUT),
    ).toBeTruthy();
    expect(
      labelElement
        .querySelectorAll("p")[0]
        .className.includes(DRAG_ZONE.ON_DRAG.STYLE_TEXT),
    ).toBeTruthy();
  });

  it("should show input, label styles and text default if drag over an image", () => {
    const { getByTestId, getByText } = setupRender();

    const labelElement = getByTestId("drag-file");

    // Simulate a dragOver event with the file
    fireEvent.dragOver(labelElement, { dataTransfer: { files: [file] } });
    // Expect input, label styles change and text after dragOver
    expect(getByText(DRAG_ZONE.DEFAULT.TEXT)).toBeTruthy();
    expect(
      labelElement.className.includes(DRAG_ZONE.DEFAULT.STYLE_INPUT),
    ).toBeTruthy();
    expect(
      labelElement
        .querySelectorAll("p")[0]
        .className.includes(DRAG_ZONE.DEFAULT.STYLE_TEXT),
    ).toBeTruthy();
  });
});
