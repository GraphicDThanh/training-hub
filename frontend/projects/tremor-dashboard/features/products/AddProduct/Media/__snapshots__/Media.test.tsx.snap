// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Media component Should match snapshot 1`] = `
<div>
  <div
    class="flex flex-col items-center justify-center h-[calc(100%-32px)] w-full"
  >
    <label
      class="flex flex-col items-center justify-center w-full h-full border border-gray-300 rounded-lg hover:cursor-pointer"
      data-testid="drag-file"
    >
      <div
        class="flex flex-col items-center justify-center pt-5 pb-6 mx-2"
      >
        <p
          class="text-tremor-default text-center text-gray-500 dark:text-gray-400"
        >
          Drag and drop your file here to begin uploading.
        </p>
        <p
          class="text-tremor-default text-xs text-yellow-700 dark:text-yellow-500 mt-2 px-2 text-center"
        >
          (We accept .png, .jpg, .jpeg, and .webp files, up to 32MB in size.)
        </p>
      </div>
      <input
        accept="image/*"
        class="hidden"
        data-testid="dropzone-file"
        id="dropzone-file"
        type="file"
      />
    </label>
  </div>
</div>
`;
